site/_shortcodes/Example.js

```diff
@@ -17,11 +17,186 @@
 const Prism = require('prismjs');
-const md = require('markdown-it')();
+const espree = require('espree');
+const execa = require('execa');
 
 const examples = [];
 let counter = {};
 
+class Example {
+  constructor(nunjucksEngine, context, raw) {
+    this.renderer = nunjucksEngine;
+
+    this.context = context;
+    this.raw = raw;
+
+    this.title = context.title;
+    this.slug = context.page.fileSlug;
+    counter[this.slug] = counter[this.slug] + 1 || 0;
+    this.id = `${this.slug}-${counter[this.slug]}`;
+
+    this.parsed = {
+      webComponents: this._parseHtml(),
+      react: this._parseJavaScript(),
+    }
+
+    this.previews = this._renderPreviews();
+    this.widget = this._renderWidget();
+  }
+
+  _extract(regExp, string) {
+    const match = regExp.exec(string);
+    return match && match[1] ? match[1] : null;
+  }
+
+  _parseCode(language, string) {
+    const code = {
+      all: this._extract(new RegExp('```' + language + '\n(.*?)\n```', 'gms'), string),
+    };
+
+    return code;
+  }
+
+  _parseHtml() {
+    const html = this._parseCode('html', this.raw);
+    if (!html.all) {
+      return null;
+    }
+
+    html.head = this._extract(/<head>(.*?)<\/head>/gms, html.all);
+    html.body = this._extract(/<body>(.*?)<\/body>/gms, html.all);
+
+    // Make sure head and body both don't contain any blank lines
+    // as they would create empty paragraphs in markdown
+    html.head = html.head.replace(/^\s*\n/gm, '').replace(/\s*$/, '');
+    html.body = html.body.replace(/^\s*\n/gm, '').replace(/\s*$/, '');
+
+    return html;
+  }
+
+  _parseJavaScript() {
+    const js = this._parseCode('javascript', this.raw);
+    if (!js.all) {
+      return null;
+    }
+
+    let ast = {};
+    try {
+      ast = espree.parse(js.all, {
+        ecmaVersion: 'latest',
+        sourceType: 'module',
+        ecmaFeatures: {
+          jsx: true,
+        }
+      });
+    } catch (e) {
+      console.error(`[Examples] Could not retrieve AST for ${this.fileSlug}`);
+      return null;
+    }
+
+    js.imports = ast.body.filter((node) => {
+      return node.type === 'ImportDeclaration'
+    }).map((node) => {
+      return js.all.substring(node.start, node.end);
+    });
+
+    js.app = ast.body.find((node) => {
+      return node.type === 'FunctionDeclaration'
+    });
+
+    if (!js.app) {
+      console.warn(`[Examples] No function in ${this.fileSlug}`);
+      return null;
+    }
+
+    js.appName = js.app.id.name;
+    js.app = js.all.substring(js.app.start, js.app.end);
+
+    return js;
+  }
+
+  _renderPreviews() {
+    const previews = {};
+    if (this.parsed.webComponents) {
+      previews.webComponents = this.renderer.render('site/_includes/layouts/example.njk', {
+        id: this.id,
+        html: this.parsed.webComponents,
+        title: this.title,
+      });
+    }
+
+    if (this.parsed.react) {
+      previews.react = this.renderer.render('site/_includes/files/componentExample.js.njk', {
```

Comment:
Why are the example templates in different directories?

Reply:
I wanted to make clear that this template doesn't make sense in the 11ty/HTML context by putting it in a different directory.

Reply:
Hm. Not sure if that is needed. I'd prefer to keep all templates in the same hierachy, but maybe add a comment in the template.